iceoryx_hoofs/utility/include/iox/detail/convert.inl

```diff
@@ -58,386 +60,449 @@ convert::toString(const Source& t) noexcept
+    return evaluate_return_value<signed char>(call, end_ptr, v);
+}
+
+template <typename TargetType, typename SourceType>
+inline bool convert::check_edge_case(decltype(errno) errno_cache,
+                                     const char* end_ptr,
+                                     const char* v,
+                                     const SourceType& source_val) noexcept
+{
+    return is_valid_input(end_ptr, v, source_val) && is_valid_errno(errno_cache, v)
+           && is_within_range<TargetType>(source_val);
+}
 
+template <typename TargetType, typename CallType>
+inline iox::optional<TargetType>
+convert::evaluate_return_value(CallType& call, const char* end_ptr, const char* v) noexcept
+{
     if (call.has_error())
     {
-        return false;
+        return iox::nullopt;
     }
 
-    if (call->value > std::numeric_limits<uint16_t>::max())
+    if (!check_edge_case<TargetType>(call->errnum, end_ptr, v, call->value))
     {
-        IOX_LOG(DEBUG, call->value << " too large, uint16_t overflow");
-        return false;
+        return iox::nullopt;
     }
 
-    dest = static_cast<uint16_t>(call->value);
-    return true;
+    return iox::optional<TargetType>(static_cast<TargetType>(call->value));
 }
 
-template <>
-inline bool convert::fromString<uint8_t>(const char* v, uint8_t& dest) noexcept
+template <typename SourceType>
+inline bool convert::is_valid_input(const char* end_ptr, const char* v, const SourceType& source_val) noexcept
 {
-    if (!stringIsNumberWithErrorMessage(v, NumberType::UNSIGNED_INTEGER))
-    {
-        return false;
-    }
-
-    auto call = IOX_POSIX_CALL(strtoul)(v, nullptr, STRTOULL_BASE).failureReturnValue(ULONG_MAX).evaluate();
-
-    if (call.has_error())
+    // invalid string
+    if (v == end_ptr && source_val == 0)
     {
+        IOX_LOG(DEBUG, "invalid input");
         return false;
     }
 
-    if (call->value > std::numeric_limits<uint8_t>::max())
+    // end_ptr is not '\0' which means conversion failure at end_ptr
+    if (end_ptr != nullptr && v != end_ptr && *end_ptr != '\0')
     {
-        IOX_LOG(DEBUG, call->value << " too large, uint8_t overflow");
+        IOX_LOG(DEBUG, "conversion failed at " << end_ptr - v << " : " << *end_ptr);
         return false;
     }
 
-    dest = static_cast<uint8_t>(call->value);
     return true;
 }
 
-template <>
-inline bool convert::fromString<int64_t>(const char* v, int64_t& dest) noexcept
+template <typename TargetType, typename SourceType>
+inline bool convert::is_within_range(const SourceType& source_val) noexcept
 {
-    if (!stringIsNumberWithErrorMessage(v, NumberType::INTEGER))
+    if constexpr (std::is_arithmetic_v<TargetType> == false)
     {
-        return false;
+        return true;
+    }
+    // is_arithmetic_v
+    if constexpr (std::is_floating_point_v<SourceType>)
+    {
+        // special cases for floating point
+        if (std::isnan(source_val))
+        {
+            return !is_signaling_nan(source_val);
```

Comment:
This can become just `return true` and can be combined with the statement below.